contracts/RelayHub.sol

```diff
@@ -266,80 +288,94 @@ contract RelayHub is IRelayHub {
         }
 
         // From this point on, this transaction will not revert nor run out of gas, and the recipient will be charged
         // for the gas spent.
 
         // The sender's nonce is advanced to prevent transaction replays.
-        nonces[from]++;
+        nonces[metaTx.from]++;
+
+        uint256 preChecksGas = initialGas - gasleft();
 
         // Calls to the recipient are performed atomically inside an inner transaction which may revert in case of
         // errors in the recipient. In either case (revert or regular execution) the return data encodes the
         // RelayCallStatus value.
-        (, bytes memory relayCallStatus) = address(this).call(abi.encodeWithSelector(this.recipientCallsAtomic.selector, from, recipient, msg.sender, encodedFunction, transactionFee, gasLimit, initialGas));
+        (, bytes memory relayCallStatus) = address(this).call(abi.encodeWithSelector(this.recipientCallsAtomic.selector, metaTx, preChecksGas));
         RelayCallStatus status = abi.decode(relayCallStatus, (RelayCallStatus));
 
+        // We know perform the actual charge calculation, based on the measured gas used
+        uint256 charge = calculateCharge(
+            getChargeableGas(initialGas - gasleft(), false),
+            metaTx.gasPrice,
+            metaTx.transactionFee
+        );
+
         // Regardless of the outcome of the relayed transaction, the recipient is now charged.
-        uint256 charge = getChargedAmount(gasOverhead + initialGas - gasleft(), gasPrice, transactionFee);
 
         // We've already checked that the recipient has enough balance to pay for the relayed transaction, this is only
         // a sanity check to prevent overflows in case of bugs.
-        require(balances[recipient] >= charge, "Should not get here");
-        balances[recipient] -= charge;
-        balances[relays[msg.sender].owner] += charge;
+        require(balances[metaTx.recipient] >= charge, "Should not get here");
+        balances[metaTx.recipient] -= charge;
+        balances[relays[metaTx.relay].owner] += charge;
 
-        emit TransactionRelayed(msg.sender, from, recipient, functionSelector, status, charge);
-    }
-
-    function getChargedAmount(uint256 gas, uint256 gasPrice, uint256 fee) private pure returns (uint256) {
-        // The fee is expressed as a percentage. E.g. a value of 40 stands for a
-        // 40% fee, so the recipient will be charged for 1.4 times the spent
-        // amount.
-        return (gas * gasPrice * (100 + fee)) / 100;
+        emit TransactionRelayed(metaTx.relay, metaTx.from, metaTx.recipient, metaTx.functionSelector, status, charge);
     }
 
     function recipientCallsAtomic(
-        address from,
-        address recipient,
-        address relayAddr,
-        bytes calldata encodedFunction,
-        uint256 transactionFee,
-        uint256 gasLimit,
-        uint256 initialGas
+        MetaTxData memory metaTx,
+        uint256 preChecksGas
     )
-    external
+    public
```

Comment:
Shouldn't we instead change almost all other functions to external instead of public? this (and most other functions) are not going to be called internally anyway.

Reply:
That's not a bad idea, but I didn't want to include more changes that needed on this PR, considering its scope grew due to unrelated reasons.

Unfortunately, `recipientCallsAtomic` _has_ to remain `public` due to a bug in the Solidity compiler, which causes it to disallow copying structs from `calldata` into `memory`. The only way to have the struct be in `memory` is by having the function be `public` (`external` requires a `calldata` location).

I've spoken with Chris (lead Solidity dev) about this, and they don't have an estimate for when a fix will be released.